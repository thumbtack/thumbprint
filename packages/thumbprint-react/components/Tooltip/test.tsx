import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import Tooltip from './index';

jest.useFakeTimers();

// https://github.com/thumbtack/thumbprint/issues/72
// https://github.com/FezVrasta/popper.js/issues/478#issuecomment-341494703
jest.mock('popper.js', () => {
    const PopperJS = jest.requireActual('popper.js');

    class Popper {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        static placements: any;

        constructor() {
            return {
                destroy: (): void => {},
                scheduleUpdate: (): void => {},
            };
        }
    }

    Popper.placements = PopperJS.placements;

    return Popper;
});

describe('Tooltip', () => {
    test('renders a closed tooltip', () => {
        const wrapper = mount(
            <Tooltip text="Goose">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('renders an open tooltip', () => {
        const wrapper = mount(
            <Tooltip text="Goose">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        const button = wrapper.find('button');
        button.simulate('focus');
        jest.runAllTimers();

        expect(wrapper).toMatchSnapshot();

        button.simulate('blur');
        jest.runAllTimers();
    });

    test('adds `zIndex`', () => {
        const wrapper = mount(
            <Tooltip text="Goose" zIndex={123}>
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        const button = wrapper.find('button');
        button.simulate('focus');
        jest.runAllTimers();

        expect(wrapper).toMatchSnapshot();

        button.simulate('blur');
        jest.runAllTimers();
    });

    test('renders an open tooltip with `bottom` placement', () => {
        const wrapper = mount(
            <Tooltip text="Goose" position="bottom">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        const button = wrapper.find('button');
        button.simulate('focus');
        jest.runAllTimers();
        expect(wrapper).toMatchSnapshot();
        button.simulate('blur');
        jest.runAllTimers();
    });

    test('renders an open tooltip with a `light` theme', () => {
        const wrapper = mount(
            <Tooltip text="Goose" theme="light">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        const button = wrapper.find('button');
        button.simulate('focus');
        jest.runAllTimers();
        expect(wrapper).toMatchSnapshot();
        button.simulate('blur');
        jest.runAllTimers();
    });

    test('renders an inline tooltip', () => {
        const wrapper = mount(
            <Tooltip text="Goose" container="inline">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        const button = wrapper.find('button');
        button.simulate('focus');
        jest.runAllTimers();
        expect(wrapper).toMatchSnapshot();
        button.simulate('blur');
        jest.runAllTimers();
    });

    test('Pressing the `Esc` key closes the tooltip', () => {
        const wrapper = mount(
            <Tooltip text="Goose" container="inline">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );

        const button = wrapper.find('button');

        expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

        button.simulate('focus');
        expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

        act(() => {
            // NOTE: it seems that Enzyme does not respond to the standard properties on
            // `KeyboardEventInit`, either `code: 'Escape'` or `key: 'Escape'`. It only responds to
            // the deprecated property `keyCode`, which is not in the type defintion. As such we
            // have to cast the type here to prevent an error.
            document.dispatchEvent(
                new KeyboardEvent('keyup', {
                    key: 'Escape',
                    code: 'Escape',
                    keyCode: 27,
                } as KeyboardEventInit),
            );
        });
        wrapper.update();

        expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
    });

    test('Events from tooltip element propagate to parent element', () => {
        const jestOnClick = jest.fn();
        const wrapper = mount(
            <Tooltip text="Goose">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={(): void => {
                            onClick();
                            jestOnClick();
                        }}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );
        expect(jestOnClick).toHaveBeenCalledTimes(0);
        wrapper.find('button').simulate('click');
        expect(jestOnClick).toHaveBeenCalledTimes(1);
    });

    test('Events from PositionedTooltip do not propagate to parent container when `container` is `body`', () => {
        const jestOnClick = jest.fn();

        const wrapper = mount(
            // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
            <div onClick={jestOnClick}>
                <Tooltip text="Goose" container="body">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>
            </div>,
        );

        expect(jestOnClick).toHaveBeenCalledTimes(0);
        wrapper.find('button').simulate('mouseenter');
        expect(jestOnClick).toHaveBeenCalledTimes(0);

        act(() => {
            jest.runAllTimers();
        });

        wrapper
            // Need to manually call update here because setState is called asynchronously in the
            // show method.
            .update()
            .find('[role="tooltip"]')
            .simulate('click');
        expect(jestOnClick).toHaveBeenCalledTimes(0);
    });

    test('Events from PositionedTooltip do propagate to parent container when `container` is `inline`', () => {
        const jestOnClick = jest.fn();

        const wrapper = mount(
            // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
            <div onClick={jestOnClick}>
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>
            </div>,
        );
        expect(jestOnClick).toHaveBeenCalledTimes(0);
        wrapper.find('button').simulate('mouseenter');
        expect(jestOnClick).toHaveBeenCalledTimes(0);

        act(() => {
            jest.runAllTimers();
        });

        wrapper
            // Need to manually call update here because setState is called asynchronously in the
            // show method.
            .update()
            .find('[role="tooltip"]')
            .simulate('click');
        expect(jestOnClick).toHaveBeenCalledTimes(1);
    });

    describe('non-touch devices', () => {
        test('`mouseenter` and `mouseleave` open and close the tooltip', () => {
            const wrapper = mount(
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            button.simulate('mouseenter');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            act(() => {
                jest.runAllTimers();
            });
            wrapper.update();
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('mouseleave');
            act(() => {
                jest.runAllTimers();
            });
            wrapper.update();
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test('`focus` and `blur` open and close the tooltip', () => {
            const wrapper = mount(
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            button.simulate('focus');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('blur');
            act(() => {
                jest.runAllTimers();
            });
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test("click doesn't do anything", () => {
            const wrapper = mount(
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            button.simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test("click on tooltip body doesn't do anything", () => {
            const wrapper = mount(
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );
            const button = wrapper.find('button');
            button.simulate('focus');
            jest.runAllTimers();

            wrapper.find('[role="tooltip"]').simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('blur');
            jest.runAllTimers();
        });
    });

    describe('touch devices', () => {
        beforeAll(() => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (global as any).window.ontouchstart = {};
        });

        afterAll(() => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            delete (global as any).window.ontouchstart;
        });

        test('`click` opens the tooltip on first click and closes on second', () => {
            const wrapper = mount(
                <Tooltip text="Goose">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            button.simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('click');
            jest.runAllTimers();
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test("focus doesn't do anything", () => {
            const wrapper = mount(
                <Tooltip text="Goose">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            button.simulate('focus');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test('`blur` closes the tooltip', () => {
            const wrapper = mount(
                <Tooltip text="Goose">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');
            button.simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('blur');
            act(() => {
                jest.runAllTimers();
            });
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test('`mouseleave` closes the tooltip after a 200ms delay', () => {
            const delayLength = 200;

            const wrapper = mount(
                <Tooltip text="Goose" closeDelayLength={delayLength}>
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
            const button = wrapper.find('button');

            button.simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('mouseleave');
            act(() => {
                jest.runAllTimers();
            });
            wrapper.update();
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test('`mouseleave` closes the tooltip immediately if `closeDelayLength` is `0`', () => {
            const delayLength = 0;

            const wrapper = mount(
                <Tooltip text="Goose" closeDelayLength={delayLength}>
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            button.simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(true);

            button.simulate('mouseleave');
            act(() => {
                jest.advanceTimersByTime(delayLength);
            });
            wrapper.update();
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test("mouseenter doesn't do anything", () => {
            const wrapper = mount(
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');

            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);

            button.simulate('mouseenter');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });

        test('`click` on PositionedTooltip would close it', () => {
            const wrapper = mount(
                <Tooltip text="Goose" container="inline">
                    {({
                        ref,
                        onMouseEnter,
                        onClick,
                        onFocus,
                        onMouseLeave,
                        onBlur,
                        ariaLabel,
                    }): JSX.Element => (
                        <button
                            ref={ref}
                            onMouseEnter={onMouseEnter}
                            onClick={onClick}
                            onFocus={onFocus}
                            onMouseLeave={onMouseLeave}
                            onBlur={onBlur}
                            aria-label={ariaLabel}
                            type="button"
                        >
                            Duck
                        </button>
                    )}
                </Tooltip>,
            );

            const button = wrapper.find('button');
            button.simulate('click');
            jest.runAllTimers();

            wrapper.find('[role="tooltip"]').simulate('click');
            expect(wrapper.find('[role="tooltip"]').exists()).toBe(false);
        });
    });

    test("text can be changed while it's open", () => {
        const wrapper = mount(
            <Tooltip text="Goose" container="inline">
                {({
                    ref,
                    onMouseEnter,
                    onClick,
                    onFocus,
                    onMouseLeave,
                    onBlur,
                    ariaLabel,
                }): JSX.Element => (
                    <button
                        ref={ref}
                        onMouseEnter={onMouseEnter}
                        onClick={onClick}
                        onFocus={onFocus}
                        onMouseLeave={onMouseLeave}
                        onBlur={onBlur}
                        aria-label={ariaLabel}
                        type="button"
                    >
                        Duck
                    </button>
                )}
            </Tooltip>,
        );
        const button = wrapper.find('button');
        button.simulate('focus');
        jest.runAllTimers();

        wrapper.setProps({ text: 'Swan' });

        expect(wrapper.find('[role="tooltip"]').text()).toContain('Swan');
    });
});
