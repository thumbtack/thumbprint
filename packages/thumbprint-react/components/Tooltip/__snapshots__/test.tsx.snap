// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip adds \`zIndex\` 1`] = `
<Tooltip
  text="Goose"
  zIndex={123}
>
  <Manager>
    <Reference>
      <InnerReference
        getReferenceRef={[Function]}
      >
        <button
          aria-label="Goose"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button"
        >
          Duck
        </button>
      </InnerReference>
    </Reference>
    <ConditionalPortal
      shouldDisplace={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="tooltip tooltipDark"
              data-test-id="tooltip"
              role="tooltip"
              style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none; z-index: 123;"
            >
              Goose
              <div
                class="nubbin nubbinDark"
              />
            </div>
          </body>
        }
      >
        <Popper
          modifiers={
            Object {
              "offset": Object {
                "offset": "0, 16px",
              },
              "preventOverflow": Object {
                "boundariesElement": "window",
              },
            }
          }
          placement="top"
          positionFixed={false}
        >
          <InnerPopper
            eventsEnabled={true}
            modifiers={
              Object {
                "offset": Object {
                  "offset": "0, 16px",
                },
                "preventOverflow": Object {
                  "boundariesElement": "window",
                },
              }
            }
            placement="top"
            positionFixed={false}
            referenceElement={
              <button
                aria-label="Goose"
                type="button"
              >
                Duck
              </button>
            }
          >
            <div
              className="tooltip tooltipDark"
              data-test-id="tooltip"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              role="tooltip"
              style={
                Object {
                  "left": 0,
                  "opacity": 0,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 0,
                  "zIndex": 123,
                }
              }
            >
              <WhenChildrenChange
                do={[Function]}
              >
                Goose
              </WhenChildrenChange>
              <div
                className="nubbin nubbinDark"
                style={Object {}}
              />
            </div>
          </InnerPopper>
        </Popper>
      </Portal>
    </ConditionalPortal>
  </Manager>
</Tooltip>
`;

exports[`Tooltip renders a closed tooltip 1`] = `
<Tooltip
  text="Goose"
>
  <Manager>
    <Reference>
      <InnerReference
        getReferenceRef={[Function]}
      >
        <button
          aria-label="Goose"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button"
        >
          Duck
        </button>
      </InnerReference>
    </Reference>
    <ConditionalPortal
      shouldDisplace={true}
    />
  </Manager>
</Tooltip>
`;

exports[`Tooltip renders an inline tooltip 1`] = `
<Tooltip
  container="inline"
  text="Goose"
>
  <Manager>
    <Reference>
      <InnerReference
        getReferenceRef={[Function]}
      >
        <button
          aria-label="Goose"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button"
        >
          Duck
        </button>
      </InnerReference>
    </Reference>
    <ConditionalPortal
      shouldDisplace={false}
    >
      <Popper
        modifiers={
          Object {
            "offset": Object {
              "offset": "0, 16px",
            },
            "preventOverflow": Object {
              "boundariesElement": "window",
            },
          }
        }
        placement="top"
        positionFixed={false}
      >
        <InnerPopper
          eventsEnabled={true}
          modifiers={
            Object {
              "offset": Object {
                "offset": "0, 16px",
              },
              "preventOverflow": Object {
                "boundariesElement": "window",
              },
            }
          }
          placement="top"
          positionFixed={false}
          referenceElement={
            <button
              aria-label="Goose"
              type="button"
            >
              Duck
            </button>
          }
        >
          <div
            className="tooltip tooltipDark"
            data-test-id="tooltip"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            role="tooltip"
            style={
              Object {
                "left": 0,
                "opacity": 0,
                "pointerEvents": "none",
                "position": "absolute",
                "top": 0,
                "zIndex": undefined,
              }
            }
          >
            <WhenChildrenChange
              do={[Function]}
            >
              Goose
            </WhenChildrenChange>
            <div
              className="nubbin nubbinDark"
              style={Object {}}
            />
          </div>
        </InnerPopper>
      </Popper>
    </ConditionalPortal>
  </Manager>
</Tooltip>
`;

exports[`Tooltip renders an open tooltip 1`] = `
<Tooltip
  text="Goose"
>
  <Manager>
    <Reference>
      <InnerReference
        getReferenceRef={[Function]}
      >
        <button
          aria-label="Goose"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button"
        >
          Duck
        </button>
      </InnerReference>
    </Reference>
    <ConditionalPortal
      shouldDisplace={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="tooltip tooltipDark"
              data-test-id="tooltip"
              role="tooltip"
              style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
            >
              Goose
              <div
                class="nubbin nubbinDark"
              />
            </div>
          </body>
        }
      >
        <Popper
          modifiers={
            Object {
              "offset": Object {
                "offset": "0, 16px",
              },
              "preventOverflow": Object {
                "boundariesElement": "window",
              },
            }
          }
          placement="top"
          positionFixed={false}
        >
          <InnerPopper
            eventsEnabled={true}
            modifiers={
              Object {
                "offset": Object {
                  "offset": "0, 16px",
                },
                "preventOverflow": Object {
                  "boundariesElement": "window",
                },
              }
            }
            placement="top"
            positionFixed={false}
            referenceElement={
              <button
                aria-label="Goose"
                type="button"
              >
                Duck
              </button>
            }
          >
            <div
              className="tooltip tooltipDark"
              data-test-id="tooltip"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              role="tooltip"
              style={
                Object {
                  "left": 0,
                  "opacity": 0,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 0,
                  "zIndex": undefined,
                }
              }
            >
              <WhenChildrenChange
                do={[Function]}
              >
                Goose
              </WhenChildrenChange>
              <div
                className="nubbin nubbinDark"
                style={Object {}}
              />
            </div>
          </InnerPopper>
        </Popper>
      </Portal>
    </ConditionalPortal>
  </Manager>
</Tooltip>
`;

exports[`Tooltip renders an open tooltip with \`bottom\` placement 1`] = `
<Tooltip
  position="bottom"
  text="Goose"
>
  <Manager>
    <Reference>
      <InnerReference
        getReferenceRef={[Function]}
      >
        <button
          aria-label="Goose"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button"
        >
          Duck
        </button>
      </InnerReference>
    </Reference>
    <ConditionalPortal
      shouldDisplace={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="tooltip tooltipDark"
              data-test-id="tooltip"
              role="tooltip"
              style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
            >
              Goose
              <div
                class="nubbin nubbinDark"
              />
            </div>
          </body>
        }
      >
        <Popper
          modifiers={
            Object {
              "offset": Object {
                "offset": "0, 16px",
              },
              "preventOverflow": Object {
                "boundariesElement": "window",
              },
            }
          }
          placement="bottom"
          positionFixed={false}
        >
          <InnerPopper
            eventsEnabled={true}
            modifiers={
              Object {
                "offset": Object {
                  "offset": "0, 16px",
                },
                "preventOverflow": Object {
                  "boundariesElement": "window",
                },
              }
            }
            placement="bottom"
            positionFixed={false}
            referenceElement={
              <button
                aria-label="Goose"
                type="button"
              >
                Duck
              </button>
            }
          >
            <div
              className="tooltip tooltipDark"
              data-test-id="tooltip"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              role="tooltip"
              style={
                Object {
                  "left": 0,
                  "opacity": 0,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 0,
                  "zIndex": undefined,
                }
              }
            >
              <WhenChildrenChange
                do={[Function]}
              >
                Goose
              </WhenChildrenChange>
              <div
                className="nubbin nubbinDark"
                style={Object {}}
              />
            </div>
          </InnerPopper>
        </Popper>
      </Portal>
    </ConditionalPortal>
  </Manager>
</Tooltip>
`;

exports[`Tooltip renders an open tooltip with a \`light\` theme 1`] = `
<Tooltip
  text="Goose"
  theme="light"
>
  <Manager>
    <Reference>
      <InnerReference
        getReferenceRef={[Function]}
      >
        <button
          aria-label="Goose"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          type="button"
        >
          Duck
        </button>
      </InnerReference>
    </Reference>
    <ConditionalPortal
      shouldDisplace={true}
    >
      <Portal
        containerInfo={
          <body>
            <div
              class="tooltip tooltipLight"
              data-test-id="tooltip"
              role="tooltip"
              style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
            >
              Goose
              <div
                class="nubbin nubbinLight"
              />
            </div>
          </body>
        }
      >
        <Popper
          modifiers={
            Object {
              "offset": Object {
                "offset": "0, 16px",
              },
              "preventOverflow": Object {
                "boundariesElement": "window",
              },
            }
          }
          placement="top"
          positionFixed={false}
        >
          <InnerPopper
            eventsEnabled={true}
            modifiers={
              Object {
                "offset": Object {
                  "offset": "0, 16px",
                },
                "preventOverflow": Object {
                  "boundariesElement": "window",
                },
              }
            }
            placement="top"
            positionFixed={false}
            referenceElement={
              <button
                aria-label="Goose"
                type="button"
              >
                Duck
              </button>
            }
          >
            <div
              className="tooltip tooltipLight"
              data-test-id="tooltip"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              role="tooltip"
              style={
                Object {
                  "left": 0,
                  "opacity": 0,
                  "pointerEvents": "none",
                  "position": "absolute",
                  "top": 0,
                  "zIndex": undefined,
                }
              }
            >
              <WhenChildrenChange
                do={[Function]}
              >
                Goose
              </WhenChildrenChange>
              <div
                className="nubbin nubbinLight"
                style={Object {}}
              />
            </div>
          </InnerPopper>
        </Popper>
      </Portal>
    </ConditionalPortal>
  </Manager>
</Tooltip>
`;
